java/algorithm/src/main/java/org.apache.arrow.algorithm.sort/VectorSorter.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.algorithm.sort;
+
+import org.apache.arrow.vector.ValueVector;
+
+/**
+ * Basic interface for sorting a vector.
+ *
+ * @param <V> the vector type.
+ */
+public interface VectorSorter<V extends ValueVector> {
```

Comment:
Since there aren't any implementations yet of in place sorting, I think I would prefer to call this OutOfPlaceSorter and create separate one for in-place sorting.  The signure of the new interface would return void.  I think this is a clearer contract than having the caller required to check for in-place vs out of place.

Reply:
Sounds reasonable.

Creating separate interfaces makes the code structure clearer. In the new commit, I have split the interface to two: InPlaceVectorSorter and OutOfPlaceVectorSorter, and they have a common super interface, VectorSorter. 

So there is a uniform way for sorting a vector. I think this is important for cases when the application does not care if the underlying sorting algorithm is in-place or out-of-place.

Reply:
Thanks for the suggestion.

I have revised the interface for the out-of-place sorter to take an output vector.

I agree that it is important to seriously consider if the sort is in-place or out-of-place. It is important enough so any use case should decide it carefully.

The problem is that, for some scenarios, the code to create the sorter and the code to use the sorter are separated, so that the code to use the sorter does not know the type of the sorter.

Let me illustrate with a simple example: suppose we have a method to sort a vector:

public IntVector sortVector(IntVector vector, VectorSorter sorter) {
  return sorter.sort(vector);
}

The above code uses a sorter, which is created elsewhere and passed to the method. The actual type of the sorter may depend on the application context (e.g. the amount of available memory, the requirement for response time, etc.). Without a uniform super interface, the code above may look like this:

public IntVector sortVector(IntVector vector, Object sorter) {
  if (sorter instanceof InPlaceSorter) {
    ((InPlaceSorter) sorter).sortInPlace(vector);
    return vector;
  } else {
    IntVector outVec = ...
    ((OutOfPlaceSorter) sorter).sortOutOfPlace(vector, outVec);
    return outVec;
  }
}

Or we can implement two methods based on sorter types:

public IntVector sortVectorInPlace(IntVector vector, InPlaceSorter sorter) {
  sorter.sort(vector);
  return vector;
}

public IntVector sortVectorOutOfPlace(IntVector vector, OutOfPlaceSorter sorter) {
  IntVector outVec = ...
  sorter.sort(vector, outVec);
  return outVec;
}

And we need a branch to call the above methods:

if (sorter instanceof InPlaceSorter) {
  return sortInPlace(vec, ((InPlaceSorter) sorter));
} else {
  return sortOutOfPlace(vec, ((OutOfPlaceSorter) sorter));
}

Anyway, the user code would be much more complicated if there would be no uniform super interface.

Reply:
@emkornfield 

After more thoughts and discussions, we have removed the super interface (VectorSorter), as you have suggested. Would you please take a look?

The reasons for us to remove the interface are as follows:

1. Although in-place and out-of-place sorters achieves similar purposes, they are different algorithms from the perspective of API design. So it is inappropriate to force them into the same class hierarchy.

2. Even if we use a unified way of calling the in-place and out-of-place sorters, we do not have a uniform way of disposing the output vector. So anyway, the caller must be aware of the sorter type.

3. For a public library, it is much easier to add an interface than to remove one. So we remove the super interface now, and if we find it necessary in the future, we may decide to add it back.